pkg/mqtt/subscription_manager.go

```diff
@@ -0,0 +1,199 @@
+)
+
+type OnClientConnectFunc func(clientID string)
+type OnClientDisconnectFunc func(clientID string)
+type OnTopicSubscribeFunc func(clientID string, topic string)
+type OnTopicUnsubscribeFunc func(clientID string, topic string)
+
+// SubscriptionManager keeps track of subscribed topics of clients of the
+// mqtt broker by subscribing to broker events.
+// This allows to get notified when a client connects or disconnects
+// or a topic is subscribed or unsubscribed.
+type SubscriptionManager struct {
+	// subscribers keeps track of the clients and their
+	// subscribed topics (and the count of subscriptions per topic).
+	subscribers     *ShrinkingMap[string, *ShrinkingMap[string, int]]
+	subscribersLock sync.RWMutex
+
+	cleanupThresholdCount int
+	cleanupThresholdRatio float32
+
+	onClientConnect    OnClientConnectFunc
+	onClientDisconnect OnClientDisconnectFunc
+	onTopicSubscribe   OnTopicSubscribeFunc
+	onTopicUnsubscribe OnTopicUnsubscribeFunc
+}
+
+func NewSubscriptionManager(
+	onClientConnect OnClientConnectFunc,
+	onClientDisconnect OnClientDisconnectFunc,
+	onTopicSubscribe OnTopicSubscribeFunc,
+	onTopicUnsubscribe OnTopicUnsubscribeFunc,
+	cleanupThresholdCount int,
+	cleanupThresholdRatio float32) *SubscriptionManager {
+
+	return &SubscriptionManager{
+		subscribers: New[string, *ShrinkingMap[string, int]](
+			WithShrinkingThresholdRatio(cleanupThresholdRatio),
+			WithShrinkingThresholdCount(cleanupThresholdCount),
+		),
+		onClientConnect:       onClientConnect,
+		onClientDisconnect:    onClientDisconnect,
+		onTopicSubscribe:      onTopicSubscribe,
+		onTopicUnsubscribe:    onTopicUnsubscribe,
+		cleanupThresholdCount: cleanupThresholdCount,
+		cleanupThresholdRatio: cleanupThresholdRatio,
+	}
+}
+
+func (s *SubscriptionManager) Connect(clientID string) {
+	s.subscribersLock.Lock()
+	defer s.subscribersLock.Unlock()
+
+	// in case the client already exists, we cleanup old subscriptions.
+	s.cleanupClientWithoutLocking(clientID)
+
+	// create a new map for the client
+	s.subscribers.Set(clientID, New[string, int](
+		WithShrinkingThresholdRatio(s.cleanupThresholdRatio),
+		WithShrinkingThresholdCount(s.cleanupThresholdCount),
+	))
+
+	if s.onClientConnect != nil {
+		s.onClientConnect(clientID)
+	}
+}
+
+func (s *SubscriptionManager) Disconnect(clientID string) {
+	s.subscribersLock.Lock()
+	defer s.subscribersLock.Unlock()
+
+	// cleanup the client
+	s.cleanupClientWithoutLocking(clientID)
+
+	// send disconnect notification then delete the subscriber
+	if s.onClientDisconnect != nil {
+		s.onClientDisconnect(clientID)
+	}
+}
+
+func (s *SubscriptionManager) Subscribe(clientID string, topic string) {
+	s.subscribersLock.Lock()
+	defer s.subscribersLock.Unlock()
+
+	// check if the client is connected
+	subscribedTopics, has := s.subscribers.Get(clientID)
+	if !has {
+		return
+	}
+
+	count, has := subscribedTopics.Get(topic)
+	if has {
+		subscribedTopics.Set(topic, count+1)
+	} else {
+		subscribedTopics.Set(topic, 1)
+	}
```

Comment:
if duplicate topic subscription happened, it must be something wrong with the mqtt client. just return to ignore or add a warning log.
```suggestion
	count, has := subscribedTopics.Get(topic)
	if has {
		return
	} else {
		subscribedTopics.Set(topic, true)
	}
```

Reply:
I exactly thought the same, but then I decided to better leave this logic with the counting.
I know this way the count should always be 0 or 1, but exactly in the case the MQTT Broker has some weird issues with duplicated topics (race conditions etc), we are safe against that.
If we fail to correctly handle the subscriptions, then we end up subscribing to a lot of INX topics. Better be safe than sorry in my opinion.